src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -90,6 +92,10 @@ public Renderer(IServiceProvider serviceProvider, ILoggerFactory loggerFactory,
         _logger = loggerFactory.CreateLogger("Microsoft.AspNetCore.Components.RenderTree.Renderer");
         _componentFactory = new ComponentFactory(componentActivator, this);
 
+        // TODO register RenderingMetrics as singleton in DI
+        var meterFactory = serviceProvider.GetService<IMeterFactory>();
+        _renderingMetrics = meterFactory != null ? new RenderingMetrics(meterFactory) : null;
```

Comment:
Is there a case where `IMeterFactory` is not ever on DI?

Reply:
For the unit tests we would simply update them. For production workloads our hosts should just call `AddMetrics` and rely on it being there. I don't think this dependency should be optional.

Reply:
What about WASM, do you think it should always have metrics in DI ? That would make impossible to disable/trim metrics for Blazor

Reply:
I think that's fine, if we are concerned, we can look at the size before and after the change to understand the delta. In any case we could put it behind and app compat switch so that it gets trimmed by default if not enabled?

Reply:
There is a fair bit of code involved with metrics. I think you'll want a way to toggle it on and off in WASM.